File: problems/p74/p74.go
// https://leetcode.com/problems/search-a-2d-matrix/description/

package p74

func searchMatrix(matrix [][]int, target int) bool {
	var m, n = len(matrix), len(matrix[0])
	var left, right = 0, m * n
	for left < right {
		var mid = (left + right) >> 1
		var val = matrix[mid/n][mid%n]
		if val == target {
			return true
		} else if val > target {
			right = mid
		} else {
			left = mid + 1
		}

	}
	return false
}

File: problems/p74/p74.py
from typing import List
import unittest


class Solution:
    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
        m, n = len(matrix), len(matrix[0])
        left, right = 0, m*n
        while left < right:
            mid = int((left+right)/2)
            val = matrix[int(mid/n)][mid % n]
            if val == target:
                return True
            elif val > target:
                right = mid
            else:
                left = mid + 1

        return False


class Test(unittest.TestCase):
    def setUp(self) -> None:
        self.s = Solution()

    def test_example1(self) -> None:
        self.assertEqual(
            True,
            self.s.searchMatrix(
                [[1, 3, 5, 7], [10, 11, 16, 20], [23, 30, 34, 60]], 3
            )
        )

    def test_example2(self) -> None:
        self.assertEqual(
            False,
            self.s.searchMatrix(
                [[1, 3, 5, 7], [10, 11, 16, 20], [23, 30, 34, 60]], 13
            )
        )


if __name__ == "__main__":
    unittest.main("p74")
